nemo/collections/nlp/modules/common/megatron/attention.py

```diff
@@ -911,3 +930,84 @@ def forward(
         context_layer = context_layer.view(*new_context_layer_shape)
 
         return context_layer
+
+    def reset_fa_causal(self, query_length, key_length, causal):
+        if query_length != key_length:
+            if query_length == 1:
+                return False
+            raise NotImplementedError(
+                "Flash attention does not support query and key with different number of tokens, unless number of query tokens is 1."
+            )
+        return causal
+
+    def flash_attention(self, query_layer, key_layer, value_layer, attention_mask):
+        batch_size, seqlen, nheads, _ = query_layer.shape
+
+        # True: attend / False: not attend
+        if attention_mask is None:
+            attention_mask_q = torch.ones(batch_size, query_layer.shape[1], device=query_layer.device).bool()
+            attention_mask_kv = torch.ones(batch_size, key_layer.shape[1], device=query_layer.device).bool()
+        elif len(attention_mask.shape) == 4:
+            # [b, 1, sq, sk] -> [b, sq] / [b, sk]
+            attention_mask_q = torch.any(torch.eq(attention_mask, False), dim=3).squeeze(1)
+            attention_mask_kv = torch.any(torch.eq(attention_mask, False), dim=2).squeeze(1)
+        else:
+            assert len(attention_mask.shape) == 2
+            attention_mask_q = attention_mask
+            attention_mask_kv = attention_mask
+
+        q, indices_q, cu_seqlens_q, max_seqlen_q = unpad_input(query_layer, attention_mask_q)
+        k, _, cu_seqlens_k, max_seqlen_k = unpad_input(key_layer, attention_mask_kv)
+        v, _, _, _ = unpad_input(value_layer, attention_mask_kv)
+        causal = self.reset_fa_causal(
+            query_layer.shape[1], key_layer.shape[1], self.attn_mask_type == AttnMaskType.causal
+        )
```

Comment:
I really would prefer a wrapper class with all of this logic, then if `use_flash_attention` we use the wrapper class. 

Can we do a refactor for this?

Reply:
I think `CoreAttention` is already the lowest module to compute attention between q, k, v. Flash attention is used to replace the operation of `softmax`, `add_attention_bias`, and `multiply value`. Therefore, I refactor with different attention function `torch_attention()` and `flash_attention()`.